controller/lib/hal/hal_stm32.cpp

```diff
@@ -0,0 +1,856 @@
+/* Copyright 2020, RespiraWorks
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#if defined(BARE_STM32)
+
+#include "hal.h"
+#include "stm32.h"
+
+// This is the main stack 
+uint32_t mainStack[ 500 ];
+
+// local data 
+static volatile uint32_t msCount;
+
+// local static functions.  I don't want to add any private 
+// functions to the Hal class to avoid complexity with other
+// builds
+static void InitGPIO( void );
+static void InitADC( void );
+static void InitSysTimer( void );
+static void InitPwmOut( void );
+static void InitUARTs( void );
+static void EnableClock( uint32_t base );
+static void EnableInterrupt( int addr, int pri );
+static void DisableInterrupt( int addr );
+static void Timer6ISR( void );
+static void UART3_ISR( void );
+
+// For now, the main function in main.cpp is called setup
+// rather then main.  If we adopt this HAL then we can 
+// just rename it main and get rid of the following function.
+extern void setup( void );
+int main( void ){
+   setup();
+}
+
+
+// This function is called from the libc initialization code 
+// before any static constructors are called.  We do some basic
+// chip initialization here.
+extern "C" void _init( void ){
+
+   // Enable the FPU
```

Comment:
Here and everywhere below (not just this function but also the other ones), please link to sections of the manual corresponding to what's going on whenever it's something hardware-specific.

The criterion would be "a reasonable person who had no prior experience with STM32, when following your reference, should be able to confirm that the code is using the device correctly within a few minutes".

I realize this is kind of a lot to ask, but as a reviewer I need to understand what's going on, and so will future readers.

For example, I tried to find out whether these two lines of code are correct, but couldn't quickly find the section of the [manual](https://www.st.com/resource/en/reference_manual/dm00151940-stm32l41xxx42xxx43xxx44xxx45xxx46xxx-advanced-armbased-32bit-mcus-stmicroelectronics.pdf) saying how to enable the FPU. The manual mentions the word "FPU" about 19 times and none of those places say how to enable it, or even that it _needs_ to be enabled.

Reply:
Fair point.  I'll add more documentation and point to specific chapters of the reference manual.
FYI - Enabling the FPU is low level ARM stuff so that isn't in the reference manual it's in the programmer's manual.  I'll add a link to that document to the comments at the top of the file as well.

Reply:
Should we add a directory to our repo to hold these important pdf files?  I added URLs to the comments, but ST might decide to move them around in the future.